kv/task.go

```diff
@@ -417,6 +413,45 @@ func (s *Service) findTasksByOrg(ctx context.Context, tx Tx, filter influxdb.Tas
 	return ts, len(ts), err
 }
 
+type taskMatchFn func(*influxdb.Task) bool
+
+// newTaskMatchFn returns a function for validating
+// a task matches the filter. Will return nil if
+// the filter should match all tasks.
+func newTaskMatchFn(f influxdb.TaskFilter, org *influxdb.Organization) func(t *influxdb.Task) bool {
+	var fn taskMatchFn
+
+	if org != nil {
+		expected := org.ID
+		prevFn := fn
+		fn = func(t *influxdb.Task) bool {
+			res := prevFn == nil || prevFn(t)
+			return res && expected == t.OrganizationID
+		}
+	}
+
+	if f.Type != nil {
+		expected := *f.Type
+		prevFn := fn
+		fn = func(t *influxdb.Task) bool {
+			res := prevFn == nil || prevFn(t)
+			return res &&
+				((expected == influxdb.TaskSystemType && (t.Type == influxdb.TaskSystemType || t.Type == "")) || expected == t.Type)
+		}
+	}
+
+	if f.Name != nil {
+		expected := *f.Name
+		prevFn := fn
+		fn = func(t *influxdb.Task) bool {
+			res := prevFn == nil || prevFn(t)
+			return res && (expected == t.Name)
+		}
+	}
+
+	return fn
```

Comment:
This is a take it or leave it nitpick:
I find this function a little hard to grok because of the way it re-composes the closures as you go down. Along with the occassional long boolean expression.
My suggestion would be to do the boring thing and just return a single closure which exits early on any non-filter matching conditions. Something like:
```go
return func(t *influxdb.Task) bool {
    // given a org filter which doesn't match the tasks organization ID
    if org != nil && org.ID != t.OrganizationID {
       return false
    }

    // given a type filter
    if f.Type != nil {
        // legacy case - from when tasks didn't have a type
        // given the filter is not for the empty type and not for the system type
        if t.Type == "" && !(*f.Type == "" || *f.Type == influxdb.TaskSystemType) {
            return false
        }

        if t.Type != *f.Type {
            return false
        }
    }
    // ...
}
```

Reply:
I tend to favor a functional approach, as they often produce efficient predicates and in many cases like this one, we can avoid conditional branches.

I will write some tests for the function, to ensure correctness.